atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/schema/metadata/SchemaMetadataServiceImpl.java

```diff
@@ -0,0 +1,155 @@
+    @VisibleForTesting
+    class InitializingWrapper extends AsyncInitializer implements AutoDelegate_SchemaMetadataService {
+        @Override
+        public SchemaMetadataService delegate() {
+            checkInitialized();
+            return SchemaMetadataServiceImpl.this;
+        }
+
+        @Override
+        protected void tryInitialize() {
+            SchemaMetadataServiceImpl.this.tryInitialize();
+        }
+
+        @Override
+        protected String getInitializingClassName() {
+            return SchemaMetadataService.class.getSimpleName();
+        }
+    }
+
+    private static final Logger log = LoggerFactory.getLogger(SchemaMetadataServiceImpl.class);
+
+    private static final long QUERY_TIMESTAMP = 1L;
+    private static final byte[] COLUMN_NAME = PtBytes.toBytes("m");
+
+    private final KeyValueService keyValueService;
+    private final InitializingWrapper wrapper = new InitializingWrapper();
+
+    private SchemaMetadataServiceImpl(KeyValueService keyValueService) {
+        this.keyValueService = keyValueService;
+    }
+
+    public static SchemaMetadataService create(KeyValueService keyValueService, boolean initializeAsync) {
+        SchemaMetadataServiceImpl metadataService = new SchemaMetadataServiceImpl(keyValueService);
+        metadataService.wrapper.initialize(initializeAsync);
+        return metadataService.wrapper.isInitialized() ? metadataService : metadataService.wrapper;
+    }
+
+    protected void initialize(boolean asyncInitialize) {
+        wrapper.initialize(asyncInitialize);
+    }
+
+    private void tryInitialize() {
+        keyValueService.createTable(
+                AtlasDbConstants.DEFAULT_SCHEMA_METADATA_TABLE, AtlasDbConstants.GENERIC_TABLE_METADATA);
+    }
+
+    @Override
+    public Optional<SchemaMetadata> loadSchemaMetadata(String schemaName) {
+        return loadMetadataCellFromKeyValueService(schemaName)
+                .map(SchemaMetadata.HYDRATOR::hydrateFromBytes);
+    }
+
+    @Override
+    public void putSchemaMetadata(String schemaName, SchemaMetadata schemaMetadata) {
+        byte[] serializedMetadata = schemaMetadata.persistToBytes();
+        while (!Thread.currentThread().isInterrupted()) {
+            Optional<byte[]> existingMetadata = loadMetadataCellFromKeyValueService(schemaName);
+            CheckAndSetRequest request = existingMetadata.map(existingData -> CheckAndSetRequest.singleCell(
+                    AtlasDbConstants.DEFAULT_SCHEMA_METADATA_TABLE,
+                    createCellForGivenSchemaName(schemaName),
+                    existingData,
+                    serializedMetadata))
+                    .orElse(CheckAndSetRequest.newCell(
+                            AtlasDbConstants.DEFAULT_SCHEMA_METADATA_TABLE,
+                            createCellForGivenSchemaName(schemaName),
+                            serializedMetadata));
+            try {
+                keyValueService.checkAndSet(request);
```

Comment:
Should log success here, because having a lot of "Retrying" statements followed by nothing from the class would be confusing.

Reply:
Agreed, added the logging statement.